File: Online Judges/LeetCode/June LeetCoding Challenge/Week-4/Reconstruct Itinerary.cpp
class Solution
{
public:
    map<string, multiset<string>> mp;
    vector<string> vec;
    void dfs(string u)
    {
        while (mp[u].size() != 0)
        {
            string v = *mp[u].begin();
            mp[u].erase(mp[u].begin());
            dfs(v);
        }
        vec.push_back(u);
    }
    vector<string> findItinerary(vector<vector<string>> &tickets)
    {
        for (auto ticket : tickets)
        {
            mp[ticket[0]].insert(ticket[1]);
        }
        dfs("JFK");
        return vector<string>(vec.rbegin(), vec.rend());
    }
};

File: Online Judges/LeetCode/June LeetCoding Challenge/Week-4/Reconstruct Itinerary.py
class Solution:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        graph = collections.defaultdict(list)

        for u, v in tickets:
            graph[u].append(v)

        for u in graph.keys():
            graph[u].sort(reverse=True)

        stack = []
        stack.append("JFK")
        res = []

        while stack:
            u = stack[-1]
            while graph[u]:
                stack.append(graph[u].pop())
                u = stack[-1]
            res.append(stack.pop())
        return res[::-1]
